In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/rem-and-nrem-sleep-classification/REM_NREM.csv


# Loading dataset

In [2]:
df = pd.read_csv("/kaggle/input/rem-and-nrem-sleep-classification/REM_NREM.csv")

In [3]:
df.columns

Index(['Sleep_Stage', 'MeanP_Alpha_F4', 'MedianF_Alpha_F4', 'MeanF_Alpha_F4',
       'Spectral Edge_Alpha_F4', 'PeakF_Alpha_F4', 'MeanP_Beta_F4',
       'MedianF_Beta_F4', 'MeanF_Beta_F4', 'Spectral Edge_Beta_F4',
       'PeakF_Beta_F4', 'MeanP_Theta_F4', 'MedianF_Theta_F4', 'MeanF_Theta_F4',
       'Spectral Edge_Theta_F4', 'PeakF_Theta_F4', 'MeanP_Delta_F4',
       'MedianF_Delta_F4', 'MeanF_Delta_F4', 'Spectral Edge_Delta_F4',
       'PeakF_Delta_F4', 'MeanP_Gamma_F4', 'MedianF_Gamma_F4',
       'MeanF_Gamma_F4', 'Spectral Edge_Gamma_F4', 'PeakF_Gamma_F4',
       'MeanP_Alpha_C4', 'MedianF_Alpha_C4', 'MeanF_Alpha_C4',
       'Spectral Edge_Alpha_C4', 'PeakF_Alpha_C4', 'MeanP_Beta_C4',
       'MedianF_Beta_C4', 'MeanF_Beta_C4', 'Spectral Edge_Beta_C4',
       'PeakF_Beta_C4', 'MeanP_Theta_C4', 'MedianF_Theta_C4', 'MeanF_Theta_C4',
       'Spectral Edge_Theta_C4', 'PeakF_Theta_C4', 'MeanP_Delta_C4',
       'MedianF_Delta_C4', 'MeanF_Delta_C4', 'Spectral Edge_Delta_C4',
       'PeakF_D

In [4]:
df.head()

Sleep_Stage  MeanP_Alpha_F4  MedianF_Alpha_F4  MeanF_Alpha_F4  \
0        NREM         0.00034          10.11512        17.85756   
1        NREM         0.00035          10.73951        17.60780   
2        NREM         0.00035           9.36585        17.60780   
3        NREM         0.00050          10.36488        16.60878   
4        NREM         0.00072           9.36585        15.98439   

   Spectral Edge_Alpha_F4  PeakF_Alpha_F4  MeanP_Beta_F4  MedianF_Beta_F4  \
0                17.48293        10.24000        0.00047         17.73268   
1                17.10829         9.11610        0.00048         16.85854   
2                16.98341         9.24098        0.00042         17.35805   
3                15.85951         7.99220        0.00053         15.48488   
4                15.48488         7.99220        0.00067         15.11024   

   MeanF_Beta_F4  Spectral Edge_Beta_F4  ...  MeanP_Delta_O2  \
0       25.35024               27.47317  ...         0.00112   
1       25.22537               26.97366  ...         0.00070   
2       26.09951               28.47220  ...         0.00077   
3       24.10146               25.72488  ...         0.00130   
4       22.22829               23.22732  ...         0.00123   

   MedianF_Delta_O2  MeanF_Delta_O2  Spectral Edge_Delta_O2  PeakF_Delta_O2  \
0           0.74927         6.36878                 2.87220         0.62439   
1           0.99902         6.86829                 3.62146         0.74927   
2           1.49854         6.49366                 4.12098         0.74927   
3           1.87317         6.11902                 3.74634         0.37463   
4           1.49854         6.61854                 3.74634         0.74927   

   MeanP_Gamma_O2  MedianF_Gamma_O2  MeanF_Gamma_O2  Spectral Edge_Gamma_O2  \
0        0.000076          32.71805        43.83220                49.95122   
1        0.000072          32.09366        44.33171                49.95122   
2        0.000069          33.09268        43.95707                49.95122   
3        0.000070          31.59415        42.83317                49.95122   
4        0.000060          29.47122        45.08098                49.95122   

   PeakF_Gamma_O2  
0        49.95122  
1        49.95122  
2        49.95122  
3        49.95122  
4        49.95122  

[5 rows x 76 columns]

## Label Encoding The Target Feature

In [5]:
from sklearn.preprocessing import LabelEncoder

label_encoder = LabelEncoder()

df['Sleep_Stage'] = label_encoder.fit_transform(df['Sleep_Stage'])

/opt/conda/lib/python3.10/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.5
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [6]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [7]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Flatten,Dense,Dropout,BatchNormalization,Conv1D,MaxPool1D
from tensorflow.keras.preprocessing.image import ImageDataGenerator

from tensorflow.keras.optimizers import Adam

In [8]:
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))

Num GPUs Available:  2


In [11]:
df.head()

Sleep_Stage  MeanP_Alpha_F4  MedianF_Alpha_F4  MeanF_Alpha_F4  \
0            0         0.00034          10.11512        17.85756   
1            0         0.00035          10.73951        17.60780   
2            0         0.00035           9.36585        17.60780   
3            0         0.00050          10.36488        16.60878   
4            0         0.00072           9.36585        15.98439   

   Spectral Edge_Alpha_F4  PeakF_Alpha_F4  MeanP_Beta_F4  MedianF_Beta_F4  \
0                17.48293        10.24000        0.00047         17.73268   
1                17.10829         9.11610        0.00048         16.85854   
2                16.98341         9.24098        0.00042         17.35805   
3                15.85951         7.99220        0.00053         15.48488   
4                15.48488         7.99220        0.00067         15.11024   

   MeanF_Beta_F4  Spectral Edge_Beta_F4  ...  MeanP_Delta_O2  \
0       25.35024               27.47317  ...         0.00112   
1       25.22537               26.97366  ...         0.00070   
2       26.09951               28.47220  ...         0.00077   
3       24.10146               25.72488  ...         0.00130   
4       22.22829               23.22732  ...         0.00123   

   MedianF_Delta_O2  MeanF_Delta_O2  Spectral Edge_Delta_O2  PeakF_Delta_O2  \
0           0.74927         6.36878                 2.87220         0.62439   
1           0.99902         6.86829                 3.62146         0.74927   
2           1.49854         6.49366                 4.12098         0.74927   
3           1.87317         6.11902                 3.74634         0.37463   
4           1.49854         6.61854                 3.74634         0.74927   

   MeanP_Gamma_O2  MedianF_Gamma_O2  MeanF_Gamma_O2  Spectral Edge_Gamma_O2  \
0        0.000076          32.71805        43.83220                49.95122   
1        0.000072          32.09366        44.33171                49.95122   
2        0.000069          33.09268        43.95707                49.95122   
3        0.000070          31.59415        42.83317                49.95122   
4        0.000060          29.47122        45.08098                49.95122   

   PeakF_Gamma_O2  
0        49.95122  
1        49.95122  
2        49.95122  
3        49.95122  
4        49.95122  

[5 rows x 76 columns]

## Train-Test Split

In [9]:
y = df['Sleep_Stage']
X = df.drop(columns=['Sleep_Stage'],axis=1)

In [10]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

## Standard Scaling

In [11]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [12]:
print(X_train.shape)
print(X_test.shape)

(71276, 75)
(17820, 75)


In [13]:
X_train.shape[0]

71276

## Reshape the train-test for CNN

In [14]:
X_train = X_train.reshape(X_train.shape[0], X_train.shape[1], 1)
X_test = X_test.reshape(X_test.shape[0], X_test.shape[1], 1)

In [15]:
from tensorflow.keras.regularizers import l2

# Creating the CNN model

In [46]:
from keras.models import Sequential
from keras.layers import Conv1D, BatchNormalization, Dropout, Flatten, Dense

model = Sequential()
model.add(Conv1D(filters=64, kernel_size=3, activation='relu', input_shape=(X_train.shape[1], 1)))
model.add(BatchNormalization())
model.add(Dropout(0.2))

model.add(Conv1D(filters=128, kernel_size=3, activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(0.5))

model.add(Conv1D(filters=256, kernel_size=3, activation='relu'))
model.add(BatchNormalization())
model.add(Dropout(0.5))

model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))

model.add(Dense(1, activation='sigmoid'))

optimizer = tf.keras.optimizers.Adam(learning_rate=0.001)  # Adjust the optimizer as needed

In [47]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv1d_3 (Conv1D)           (None, 73, 64)            256       
                                                                 
 batch_normalization_3 (Batc  (None, 73, 64)           256       
 hNormalization)                                                 
                                                                 
 dropout_4 (Dropout)         (None, 73, 64)            0         
                                                                 
 conv1d_4 (Conv1D)           (None, 71, 128)           24704     
                                                                 
 batch_normalization_4 (Batc  (None, 71, 128)          512       
 hNormalization)                                                 
                                                                 
 dropout_5 (Dropout)         (None, 71, 128)          

In [48]:
model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])

In [51]:
history = model.fit(X_train,y_train,epochs=30,validation_data=(X_test,y_test),verbose=1)

Epoch 1/30
2228/2228 [==============================] - 18s 8ms/step - loss: 0.2186 - accuracy: 0.9050 - val_loss: 0.2059 - val_accuracy: 0.9219
Epoch 2/30
2228/2228 [==============================] - 17s 8ms/step - loss: 0.2150 - accuracy: 0.9059 - val_loss: 0.2011 - val_accuracy: 0.9206
Epoch 3/30
2228/2228 [==============================] - 17s 7ms/step - loss: 0.2146 - accuracy: 0.9068 - val_loss: 0.1917 - val_accuracy: 0.9214
Epoch 4/30
2228/2228 [==============================] - 18s 8ms/step - loss: 0.2131 - accuracy: 0.9079 - val_loss: 0.2028 - val_accuracy: 0.9202
Epoch 5/30
2228/2228 [==============================] - 17s 8ms/step - loss: 0.2126 - accuracy: 0.9067 - val_loss: 0.1913 - val_accuracy: 0.9219
Epoch 6/30
2228/2228 [==============================] - 16s 7ms/step - loss: 0.2111 - accuracy: 0.9089 - val_loss: 0.2025 - val_accuracy: 0.9184
Epoch 7/30
2228/2228 [==============================] - 18s 8ms/step - loss: 0.2094 - accuracy: 0.9092 - val_loss: 0.2012 - val_ac

In [52]:
class_acc = model.evaluate(X_test, y_test, verbose=0)[1]
print("Test Accuracy: {:.4f}%".format(class_acc * 100))

Test Accuracy: 92.4747%


## Accuracy for each class (CNN)

In [53]:
from sklearn.metrics import confusion_matrix

y_pred = model.predict(X_test)
y_pred_binary = (y_pred > 0.5).astype(int)

confusion = confusion_matrix(y_test, y_pred_binary)

tn, fp, fn, tp = confusion.ravel()

class_0_accuracy = tp / (tp + fn)
class_1_accuracy = tn / (tn + fp)

print(f"Class 0 Accuracy: {class_0_accuracy * 100:.2f}%")
print(f"Class 1 Accuracy: {class_1_accuracy * 100:.2f}%")

557/557 [==============================] - 1s 2ms/step
Class 0 Accuracy: 70.35%
Class 1 Accuracy: 97.25%


# Random Forest Classifier

In [31]:
from sklearn.preprocessing import LabelEncoder
df = pd.read_csv("/kaggle/input/rem-and-nrem-sleep-classification/REM_NREM.csv")
label_encoder = LabelEncoder()
df['Sleep_Stage'] = label_encoder.fit_transform(df['Sleep_Stage'])
y = df['Sleep_Stage']
X = df.drop(columns=['Sleep_Stage'],axis=1)
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## Running RF with k-fold cross-validation

In [41]:
from sklearn.model_selection import cross_val_score, StratifiedKFold

kfold = StratifiedKFold(n_splits=10, shuffle=True, random_state=42)
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)

# Performing 10-fold cross-validation
scores = cross_val_score(rf_classifier, X_train, y_train, cv=kfold, scoring='accuracy')

print("Cross-validation scores:", scores)
print("Mean accuracy:", scores.mean())

rf_classifier.fit(X_train, y_train)
y_pred = rf_classifier.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Test set accuracy:", accuracy)

print("Classification Report:")
print(classification_report(y_test, y_pred))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Cross-validation scores: [0.92760943 0.92690797 0.92578563 0.93097643 0.93055556 0.92943322
 0.92830083 0.92338993 0.92844114 0.92858145]
Mean accuracy: 0.9279981581557747
Test set accuracy: 0.9326038159371492
Classification Report:
              precision    recall  f1-score   support

           0       0.94      0.98      0.96     14656
           1       0.88      0.72      0.79      3164

    accuracy                           0.93     17820
   macro avg       0.91      0.85      0.88     17820
weighted avg       0.93      0.93      0.93     17820

Confusion Matrix:
[[14351   305]
 [  896  2268]]


### As you can see here, mean accuracy of cross-validation is 0.927! Roughly the same as CNN.

## XGBoost Classifier

In [36]:
import xgboost as xgb

In [37]:
xgb_classifier = xgb.XGBClassifier(learning_rate=0.1, n_estimators=100, random_state=42)
xgb_classifier.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=42, ...)

In [38]:
y_pred = xgb_classifier.predict(X_test)

In [40]:
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy}')

print(classification_report(y_test, y_pred))

Accuracy: 0.9241863075196408
              precision    recall  f1-score   support

           0       0.94      0.97      0.95     14656
           1       0.83      0.72      0.77      3164

    accuracy                           0.92     17820
   macro avg       0.88      0.85      0.86     17820
weighted avg       0.92      0.92      0.92     17820



### 92.41% using XGB, almost like CNN.